### ISSUE TITLE
Proposal: Preliminary members (external IVT replacement proposal)

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#31198]
I propose adding a `[Preliminary]` attribute and analyzer which will allow certain `internal` members to be marked as `public` without gaining full protections of stable APIs for assembly versioning. These members can then be used by assemblies that are currently relying on `[InternalsVisibleTo]`.

> This library may include types and/or members which are designated as preliminary features. In the library source code, preliminary features are indicated by including the `[Preliminary]` attribute. While preliminary features are much more "flexible" during the release cycle of the library, certain rules do apply in order to ensure the stronger rules provided for stable features will not be violated by a change to a preliminary feature. The following list includes examples of these rules, but other rules may be imposed by basic logical constraints. The API elements referred to in the following list are assumed to be restricted to the publicly-exposed API of the library. The terms "member" and "members" refer to any field, method, property, or event. 
> * A member may only refer to a preliminary type in its signature if it is also marked preliminary. 
> * An interface may only include a preliminary member if it is also marked preliminary. 
> * An interface may only extend a preliminary interface if it is also marked preliminary. 
> * A class may only include a preliminary abstract member if either it is also marked preliminary, or all constructors for the class are marked internal. This restriction also applies to abstract classes which do not implement an abstract member declared in a base class. 

By default, an analyzer alongside would disallow the use of preliminary features. The diagnostic for this analyzer could be disabled or set to lower severity to allow preliminary features to be used.

**Comment by CyrusNajmabadi** [comment]
> An interface may only include a preliminary member if it is also marked preliminary.

This seems unfortunate.  Is it not possible to then prototype preliminary new interface capabilities in existing interfaces? (i.e. the 'IBlankLine' stuff i just went through)

**Comment by tmat** [comment]
I don't think this will work in practice.

C# and VB, which support analyzers, are not the only languages that can call such preliminary public APIs. Consider e.g. F# calling such API. Since F# doesn't support analyzers the user wouldn't know that they are calling preliminary API. Then they will get broken when it's changed.

Also, one can easily reference a library with preliminary APIs without referencing the analyzer.

**Comment by sharwell** [comment]
@tmat The analyzer is an aid, but the lack of an analyzer would not change the rules. The situation for those users would still be an improvement over the situation we have today by allowing us to limit the surface area exposed by what today is covered by IVTs.

**Comment by tmat** [comment]
Yes, F# and others have IVT to Roslyn internal members and we are in process of adding attributes to mark these members and an analyzer that guards changes in these members. 

That's different from this proposal though, if i understand it correctly, since those members are internal and thus we control what components have access to them. The attribute then makes it easier for us to avoid breaking these components accidentally.

The Preliminary attribute is however applied on public members. You can define whatever rules you want. Unless they are universally enforced someone will inevitably use the API in production without being warned and we will have to avoid changes to such API to not break them.

**Comment by sharwell** [comment]
> ... we are in process of adding attributes to mark these members and an analyzer that guards changes in these members ...

This doesn't address the high level goal of eliminating the IVTs. We are also considering other approaches but this issue focuses on just one.

**Comment by tmat** [comment]
There is no need to eliminate IVTs. The need is to avoid accidentally breaking the components that have the visibility. This can be achieved using InternalsUsedInAttribute.

**Comment by jasonmalinowski** [comment]
I think I'd prefer we just did the InternalsUsedInAttribute too: this seems like airing our dirty laundry, and I'm sure we'd see people taking references on Preliminary APIs without realizing it (analyzer doesn't work, they never installed it, etc.) and we just find ourselves in trouble.

**Comment by CyrusNajmabadi** [comment]
One way we handle this ourselves is to put prelim APIs in a different namespace.  i.e. 'experimental'.  It allows people to actually try things out (without needing IVT) while making it very clear this can change and should not be something you can take a hard dependency on.  These APIs are documented as being changable, and the onus is on anyone taking a dependency to have a plan on how to move forward if/when we do change things. 

I think there needs to be some sort of middle ground.  Having the only options be:

1. we make it public, but then are locked in forever, and:
2. it is internal and we almost never expose it except for a select group of IVTers

is really constraining.  It makes roslyn super conservative about ever making anything public.  And it great limits our ability to experiment and get the necessary feedback in order to actually make a great public API that can be supported for a long time.

**Comment by sharwell** [comment]
🔗 Alternative proposal in #31393
